File: backend/internal/service/platform_profile_registry.go
package service

import (
	"encoding/json"
	"strings"

	infraerrors "github.com/Wei-Shaw/sub2api/internal/pkg/errors"
)

// PlatformProfileRegistry is the long-term registry for provider/platform guide metadata.
// It is stored as one JSON setting so operators can insert/edit guide copy without
// hardcoding platform-specific instructions into the frontend.
type PlatformProfileRegistry struct {
	Version  int               `json:"version"`
	Profiles []PlatformProfile `json:"profiles"`
}

// PlatformProfile describes one platform/provider profile.
type PlatformProfile struct {
	Platform     string                `json:"platform"`
	ProviderID   string                `json:"provider_id,omitempty"`
	ProviderName string                `json:"provider_name,omitempty"`
	APIStyle     string                `json:"api_style,omitempty"`
	Guide        PlatformGuideMetadata `json:"guide"`
}

// PlatformGuideMetadata is safe to expose publicly and through provider-catalog.
type PlatformGuideMetadata struct {
	ProfileID     string                   `json:"profile_id"`
	Title         string                   `json:"title"`
	Description   string                   `json:"description"`
	Note          string                   `json:"note,omitempty"`
	DocsURL       string                   `json:"docs_url,omitempty"`
	DefaultClient string                   `json:"default_client,omitempty"`
	Clients       []PlatformGuideClient    `json:"clients,omitempty"`
	CopyBlocks    []PlatformGuideCopyBlock `json:"copy_blocks,omitempty"`
}

// PlatformGuideClient describes a user-facing client tab/target.
type PlatformGuideClient struct {
	ID    string   `json:"id"`
	Label string   `json:"label"`
	OS    []string `json:"os,omitempty"`
}

// PlatformGuideCopyBlock is a templated copyable guide block. The frontend replaces
// placeholders such as {{base_url}}, {{api_base_url}}, and {{api_key}} at render time.
type PlatformGuideCopyBlock struct {
	ID              string `json:"id"`
	ClientID        string `json:"client_id"`
	OS              string `json:"os,omitempty"`
	Path            string `json:"path"`
	Hint            string `json:"hint,omitempty"`
	Language        string `json:"language,omitempty"`
	ContentTemplate string `json:"content_template"`
}

func defaultPlatformProfileRegistry() PlatformProfileRegistry {
	return PlatformProfileRegistry{
		Version: 1,
		Profiles: []PlatformProfile{
			{
				Platform:     "openai",
				ProviderID:   "v-claw-openai",
				ProviderName: "OpenAI",
				APIStyle:     "openai-responses",
				Guide: PlatformGuideMetadata{
					ProfileID:     "openai",
					Title:         "OpenAI / Codex guide",
					Description:   "Use this API key with OpenAI-compatible clients and Codex. The values below are generated from the platform profile registry, not hardcoded per modal.",
					Note:          "For Codex, keep response storage disabled and point the provider base URL to your V-Claw endpoint.",
					DocsURL:       "https://platform.openai.com/docs",
					DefaultClient: "codex",
					Clients: []PlatformGuideClient{
						{ID: "codex", Label: "Codex CLI", OS: []string{"unix", "windows"}},
						{ID: "codex-ws", Label: "Codex CLI WebSocket", OS: []string{"unix", "windows"}},
						{ID: "claude", Label: "Claude Code compatibility", OS: []string{"unix", "cmd", "powershell"}},
					},
					CopyBlocks: []PlatformGuideCopyBlock{
						{ID: "openai-codex-config-unix", ClientID: "codex", OS: "unix", Path: "~/.codex/config.toml", Hint: "Codex config.toml", Language: "toml", ContentTemplate: `model_provider = "OpenAI"
model = "{{openai_model}}"
review_model = "{{openai_model}}"
model_reasoning_effort = "xhigh"
disable_response_storage = true
network_access = "enabled"
windows_wsl_setup_acknowledged = true

[model_providers.OpenAI]
name = "OpenAI"
base_url = "{{base_url}}"
wire_api = "responses"
requires_openai_auth = true

[features]
goals = true`},
						{ID: "openai-codex-auth-unix", ClientID: "codex", OS: "unix", Path: "~/.codex/auth.json", Language: "json", ContentTemplate: `{
  "OPENAI_API_KEY": "{{api_key}}"
}`},
						{ID: "openai-codex-config-windows", ClientID: "codex", OS: "windows", Path: "%userprofile%\\.codex\\config.toml", Hint: "Codex config.toml", Language: "toml", ContentTemplate: `model_provider = "OpenAI"
model = "{{openai_model}}"
review_model = "{{openai_model}}"
model_reasoning_effort = "xhigh"
disable_response_storage = true
network_access = "enabled"
windows_wsl_setup_acknowledged = true

[model_providers.OpenAI]
name = "OpenAI"
base_url = "{{base_url}}"
wire_api = "responses"
requires_openai_auth = true

[features]
goals = true`},
						{ID: "openai-codex-auth-windows", ClientID: "codex", OS: "windows", Path: "%userprofile%\\.codex\\auth.json", Language: "json", ContentTemplate: `{
  "OPENAI_API_KEY": "{{api_key}}"
}`},
						{ID: "openai-codex-ws-config-unix", ClientID: "codex-ws", OS: "unix", Path: "~/.codex/config.toml", Hint: "Codex WebSocket config.toml", Language: "toml", ContentTemplate: `model_provider = "OpenAI"
model = "{{openai_model}}"
review_model = "{{openai_model}}"
model_reasoning_effort = "xhigh"
disable_response_storage = true
network_access = "enabled"
windows_wsl_setup_acknowledged = true

[model_providers.OpenAI]
name = "OpenAI"
base_url = "{{base_url}}"
wire_api = "responses"
supports_websockets = true
requires_openai_auth = true

[features]
responses_websockets_v2 = true
goals = true`},
						{ID: "openai-codex-ws-auth-unix", ClientID: "codex-ws", OS: "unix", Path: "~/.codex/auth.json", Language: "json", ContentTemplate: `{
  "OPENAI_API_KEY": "{{api_key}}"
}`},
						{ID: "openai-codex-ws-config-windows", ClientID: "codex-ws", OS: "windows", Path: "%userprofile%\\.codex\\config.toml", Hint: "Codex WebSocket config.toml", Language: "toml", ContentTemplate: `model_provider = "OpenAI"
model = "{{openai_model}}"
review_model = "{{openai_model}}"
model_reasoning_effort = "xhigh"
disable_response_storage = true
network_access = "enabled"
windows_wsl_setup_acknowledged = true

[model_providers.OpenAI]
name = "OpenAI"
base_url = "{{base_url}}"
wire_api = "responses"
supports_websockets = true
requires_openai_auth = true

[features]
responses_websockets_v2 = true
goals = true`},
						{ID: "openai-codex-ws-auth-windows", ClientID: "codex-ws", OS: "windows", Path: "%userprofile%\\.codex\\auth.json", Language: "json", ContentTemplate: `{
  "OPENAI_API_KEY": "{{api_key}}"
}`},
					},
				},
			},
			{
				Platform:     "anthropic",
				ProviderID:   "v-claw-anthropic",
				ProviderName: "Anthropic",
				APIStyle:     "anthropic-messages",
				Guide: PlatformGuideMetadata{
					ProfileID:     "anthropic",
					Title:         "Anthropic / Claude Code guide",
					Description:   "Use this API key as an Anthropic-compatible token for Claude Code and compatible tools.",
					Note:          "Claude Code reads ANTHROPIC_BASE_URL and ANTHROPIC_AUTH_TOKEN from the shell or ~/.claude/settings.json.",
					DocsURL:       "https://docs.anthropic.com/claude/docs/claude-code",
					DefaultClient: "claude",
					Clients: []PlatformGuideClient{
						{ID: "claude", Label: "Claude Code", OS: []string{"unix", "cmd", "powershell"}},
						{ID: "opencode", Label: "OpenCode"},
					},
					CopyBlocks: []PlatformGuideCopyBlock{
						{ID: "anthropic-env-unix", ClientID: "claude", OS: "unix", Path: "Terminal", Language: "shell", ContentTemplate: `export ANTHROPIC_BASE_URL="{{base_url}}"
export ANTHROPIC_AUTH_TOKEN="{{api_key}}"
export CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC=1`},
						{ID: "anthropic-env-cmd", ClientID: "claude", OS: "cmd", Path: "Command Prompt", Language: "bat", ContentTemplate: `set ANTHROPIC_BASE_URL={{base_url}}
set ANTHROPIC_AUTH_TOKEN={{api_key}}
set CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC=1`},
						{ID: "anthropic-env-powershell", ClientID: "claude", OS: "powershell", Path: "PowerShell", Language: "powershell", ContentTemplate: `$env:ANTHROPIC_BASE_URL="{{base_url}}"
$env:ANTHROPIC_AUTH_TOKEN="{{api_key}}"
$env:CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC=1`},
						{ID: "anthropic-claude-settings-unix", ClientID: "claude", OS: "unix", Path: "~/.claude/settings.json", Hint: "VSCode Claude Code", Language: "json", ContentTemplate: `{
  "env": {
    "ANTHROPIC_BASE_URL": "{{base_url}}",
    "ANTHROPIC_AUTH_TOKEN": "{{api_key}}",
    "CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC": "1",
    "CLAUDE_CODE_ATTRIBUTION_HEADER": "0"
  }
}`},
						{ID: "anthropic-claude-settings-windows", ClientID: "claude", OS: "cmd", Path: "%userprofile%\\.claude\\settings.json", Hint: "VSCode Claude Code", Language: "json", ContentTemplate: `{
  "env": {
    "ANTHROPIC_BASE_URL": "{{base_url}}",
    "ANTHROPIC_AUTH_TOKEN": "{{api_key}}",
    "CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC": "1",
    "CLAUDE_CODE_ATTRIBUTION_HEADER": "0"
  }
}`},
						{ID: "anthropic-claude-settings-powershell", ClientID: "claude", OS: "powershell", Path: "%userprofile%\\.claude\\settings.json", Hint: "VSCode Claude Code", Language: "json", ContentTemplate: `{
  "env": {
    "ANTHROPIC_BASE_URL": "{{base_url}}",
    "ANTHROPIC_AUTH_TOKEN": "{{api_key}}",
    "CLAUDE_CODE_DISABLE_NONESSENTIAL_TRAFFIC": "1",
    "CLAUDE_CODE_ATTRIBUTION_HEADER": "0"
  }
}`},
					},
				},
			},
			{
				Platform:     "gemini",
				ProviderID:   "v-claw-gemini",
				ProviderName: "Google",
				APIStyle:     "google-native",
				Guide: PlatformGuideMetadata{
					ProfileID:     "gemini",
					Title:         "Gemini CLI guide",
					Description:   "Use this API key with Gemini-compatible clients. The base URL is normalized to the Gemini v1beta endpoint.",
					Note:          "Gemini CLI expects GOOGLE_GEMINI_BASE_URL, GEMINI_API_KEY, and GEMINI_MODEL.",
					DocsURL:       "https://ai.google.dev/gemini-api/docs",
					DefaultClient: "gemini",
					Clients: []PlatformGuideClient{
						{ID: "gemini", Label: "Gemini CLI", OS: []string{"unix", "cmd", "powershell"}},
						{ID: "opencode", Label: "OpenCode"},
					},
					CopyBlocks: []PlatformGuideCopyBlock{
						{ID: "gemini-env-unix", ClientID: "gemini", OS: "unix", Path: "Terminal", Language: "shell", ContentTemplate: `export GOOGLE_GEMINI_BASE_URL="{{gemini_base_url}}"
export GEMINI_API_KEY="{{api_key}}"
export GEMINI_MODEL="{{gemini_model}}"`},
						{ID: "gemini-env-cmd", ClientID: "gemini", OS: "cmd", Path: "Command Prompt", Language: "bat", ContentTemplate: `set GOOGLE_GEMINI_BASE_URL={{gemini_base_url}}
set GEMINI_API_KEY={{api_key}}
set GEMINI_MODEL={{gemini_model}}`},
						{ID: "gemini-env-powershell", ClientID: "gemini", OS: "powershell", Path: "PowerShell", Language: "powershell", ContentTemplate: `$env:GOOGLE_GEMINI_BASE_URL="{{gemini_base_url}}"
$env:GEMINI_API_KEY="{{api_key}}"
$env:GEMINI_MODEL="{{gemini_model}}"`},
					},
				},
			},
		},
	}
}

// DefaultPlatformProfileRegistryJSON returns the default insertable guide registry
// for OpenAI, Anthropic, and Gemini.
func DefaultPlatformProfileRegistryJSON() string {
	blob, _ := json.MarshalIndent(defaultPlatformProfileRegistry(), "", "  ")
	return string(blob)
}

// EffectivePlatformProfileRegistryJSON returns a normalized registry JSON string,
// falling back to the built-in defaults when the stored value is empty or invalid.
func EffectivePlatformProfileRegistryJSON(raw string) string {
	normalized, err := NormalizePlatformProfileRegistryJSON(raw)
	if err != nil {
		return DefaultPlatformProfileRegistryJSON()
	}
	return normalized
}

// NormalizePlatformProfileRegistryJSON validates and pretty-prints a registry payload.
func NormalizePlatformProfileRegistryJSON(raw string) (string, error) {
	registry, err := ParsePlatformProfileRegistry(raw)
	if err != nil {
		return "", err
	}
	blob, err := json.MarshalIndent(registry, "", "  ")
	if err != nil {
		return "", err
	}
	return string(blob), nil
}

// ParsePlatformProfileRegistry parses and validates a registry payload.
func ParsePlatformProfileRegistry(raw string) (PlatformProfileRegistry, error) {
	if strings.TrimSpace(raw) == "" {
		registry := defaultPlatformProfileRegistry()
		return registry, validatePlatformProfileRegistry(&registry)
	}
	var registry PlatformProfileRegistry
	if err := json.Unmarshal([]byte(raw), &registry); err != nil {
		return PlatformProfileRegistry{}, infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile registry must be valid JSON")
	}
	if registry.Version == 0 {
		registry.Version = 1
	}
	if err := validatePlatformProfileRegistry(&registry); err != nil {
		return PlatformProfileRegistry{}, err
	}
	return registry, nil
}

func validatePlatformProfileRegistry(registry *PlatformProfileRegistry) error {
	if registry == nil || len(registry.Profiles) == 0 {
		return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile registry must contain at least one profile")
	}
	seenPlatforms := make(map[string]struct{}, len(registry.Profiles))
	for i := range registry.Profiles {
		profile := &registry.Profiles[i]
		profile.Platform = strings.ToLower(strings.TrimSpace(profile.Platform))
		profile.ProviderID = strings.TrimSpace(profile.ProviderID)
		profile.ProviderName = strings.TrimSpace(profile.ProviderName)
		profile.APIStyle = strings.TrimSpace(profile.APIStyle)
		if profile.Platform == "" {
			return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile registry profile.platform is required")
		}
		if _, ok := seenPlatforms[profile.Platform]; ok {
			return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile registry contains duplicate platform: "+profile.Platform)
		}
		seenPlatforms[profile.Platform] = struct{}{}

		guide := &profile.Guide
		guide.ProfileID = strings.TrimSpace(guide.ProfileID)
		if guide.ProfileID == "" {
			guide.ProfileID = profile.Platform
		}
		guide.Title = strings.TrimSpace(guide.Title)
		guide.Description = strings.TrimSpace(guide.Description)
		guide.Note = strings.TrimSpace(guide.Note)
		guide.DocsURL = strings.TrimSpace(guide.DocsURL)
		guide.DefaultClient = strings.TrimSpace(guide.DefaultClient)
		if guide.Title == "" {
			return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile guide.title is required for "+profile.Platform)
		}
		if guide.Description == "" {
			return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile guide.description is required for "+profile.Platform)
		}

		seenClients := make(map[string]struct{}, len(guide.Clients))
		for ci := range guide.Clients {
			client := &guide.Clients[ci]
			client.ID = strings.TrimSpace(client.ID)
			client.Label = strings.TrimSpace(client.Label)
			if client.ID == "" || client.Label == "" {
				return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile guide clients require id and label for "+profile.Platform)
			}
			seenClients[client.ID] = struct{}{}
		}

		seenBlocks := make(map[string]struct{}, len(guide.CopyBlocks))
		for bi := range guide.CopyBlocks {
			block := &guide.CopyBlocks[bi]
			block.ID = strings.TrimSpace(block.ID)
			block.ClientID = strings.TrimSpace(block.ClientID)
			block.OS = strings.ToLower(strings.TrimSpace(block.OS))
			block.Path = strings.TrimSpace(block.Path)
			block.Hint = strings.TrimSpace(block.Hint)
			block.Language = strings.TrimSpace(block.Language)
			if block.ID == "" || block.ClientID == "" || block.Path == "" || strings.TrimSpace(block.ContentTemplate) == "" {
				return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile copy blocks require id, client_id, path, and content_template for "+profile.Platform)
			}
			if _, ok := seenBlocks[block.ID]; ok {
				return infraerrors.BadRequest("INVALID_PLATFORM_PROFILE_REGISTRY", "platform profile copy block ids must be unique for "+profile.Platform)
			}
			seenBlocks[block.ID] = struct{}{}
		}
	}
	return nil
}

// ProfileByPlatform returns a copy of the profile for platform, if present.
func (registry PlatformProfileRegistry) ProfileByPlatform(platform string) (PlatformProfile, bool) {
	platform = strings.ToLower(strings.TrimSpace(platform))
	if platform == "" {
		return PlatformProfile{}, false
	}
	for _, profile := range registry.Profiles {
		if strings.EqualFold(profile.Platform, platform) {
			return profile, true
		}
	}
	return PlatformProfile{}, false
}

// ResolvePlatformProfile returns a profile from raw registry JSON, with default
// registry fallback for empty/invalid storage.
func ResolvePlatformProfile(raw, platform string) (PlatformProfile, bool) {
	registry, err := ParsePlatformProfileRegistry(raw)
	if err != nil {
		registry = defaultPlatformProfileRegistry()
	}
	return registry.ProfileByPlatform(platform)
}

File: frontend/src/views/admin/settings/platformProfileRegistry.ts
export interface PlatformGuideCopyBlock {
  id?: string;
  client_id?: string;
  os?: string;
  path?: string;
  hint?: string;
  language?: string;
  content_template?: string;
}

export interface PlatformGuideClient {
  id?: string;
  label?: string;
  os?: string[];
}

export interface PlatformGuideMetadata {
  profile_id?: string;
  title?: string;
  description?: string;
  note?: string;
  docs_url?: string;
  default_client?: string;
  clients?: PlatformGuideClient[];
  copy_blocks?: PlatformGuideCopyBlock[];
}

export interface PlatformProfile {
  platform?: string;
  provider_id?: string;
  provider_name?: string;
  api_style?: string;
  guide?: PlatformGuideMetadata;
}

export interface PlatformProfileRegistry {
  version?: number;
  profiles?: PlatformProfile[];
  [key: string]: unknown;
}

export interface PlatformProfileRegistryValidation {
  ok: boolean;
  isEmpty: boolean;
  registry: PlatformProfileRegistry | null;
  errors: string[];
  warnings: string[];
}

export const platformProfileRegistryPlaceholders = [
  "{{base_url}}",
  "{{base_root}}",
  "{{api_base_url}}",
  "{{api_key}}",
  "{{openai_model}}",
  "{{gemini_model}}",
  "{{gemini_base_url}}",
  "{{antigravity_base_url}}",
  "{{antigravity_gemini_base_url}}",
] as const;

export const platformProfileRegistryPlaceholderSamples: Record<string, string> = {
  "{{base_url}}": "https://token.v-claw.org/v1",
  "{{base_root}}": "https://token.v-claw.org",
  "{{api_base_url}}": "https://token.v-claw.org/v1",
  "{{api_key}}": "sk-vcl...cted",
  "{{openai_model}}": "gpt-5.5",
  "{{gemini_model}}": "gemini-2.5-pro",
  "{{gemini_base_url}}": "https://token.v-claw.org/gemini/v1beta",
  "{{antigravity_base_url}}": "https://token.v-claw.org/antigravity/v1",
  "{{antigravity_gemini_base_url}}": "https://token.v-claw.org/antigravity/gemini/v1beta",
};

export const platformProfileRegistryPlaceholderSampleValues = Object.fromEntries(
  Object.entries(platformProfileRegistryPlaceholderSamples).map(([placeholder, value]) => [
    placeholder.replace(/^\{\{\s*/, "").replace(/\s*\}\}$/, ""),
    value,
  ]),
) as Record<string, string>;

export const platformProfileRegistryOSOptions = ["unix", "windows", "cmd", "powershell"] as const;

export const platformProfileRegistryLanguageOptions = [
  "toml",
  "json",
  "shell",
  "bash",
  "bat",
  "powershell",
  "text",
] as const;

export function renderPlatformProfileGuideTemplate(
  template: string,
  values: Record<string, string>,
): string {
  return template.replace(/\{\{\s*([a-zA-Z0-9_]+)\s*\}\}/g, (_match, key: string) => values[key] ?? "");
}

const defaultPlatformIDs = ["openai", "anthropic", "gemini"] as const;

export function createEmptyPlatformProfileRegistry(): PlatformProfileRegistry {
  return {
    version: 1,
    profiles: [],
  };
}

export function clonePlatformProfileRegistry(
  registry: PlatformProfileRegistry,
): PlatformProfileRegistry {
  return JSON.parse(JSON.stringify(registry)) as PlatformProfileRegistry;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function readString(value: unknown): string {
  return typeof value === "string" ? value.trim() : "";
}

function readStringArray(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value
    .filter((item): item is string => typeof item === "string")
    .map((item) => item.trim())
    .filter(Boolean);
}

function normalizePlatformID(value: string): string {
  return value.trim().toLowerCase();
}

export function validatePlatformProfileRegistryInput(
  raw: string,
): PlatformProfileRegistryValidation {
  const trimmed = raw.trim();
  if (!trimmed) {
    return {
      ok: true,
      isEmpty: true,
      registry: null,
      errors: [],
      warnings: [
        "Để trống rồi lưu sẽ để backend ghi lại registry mặc định OpenAI, Anthropic, Gemini vào DB.",
      ],
    };
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(trimmed);
  } catch (error) {
    return {
      ok: false,
      isEmpty: false,
      registry: null,
      errors: [
        `JSON không hợp lệ: ${
          error instanceof Error ? error.message : "không parse được nội dung"
        }`,
      ],
      warnings: [],
    };
  }

  const errors: string[] = [];
  const warnings: string[] = [];

  if (!isRecord(parsed)) {
    return {
      ok: false,
      isEmpty: false,
      registry: null,
      errors: ["Registry phải là một JSON object có dạng { version, profiles }."],
      warnings: [],
    };
  }

  const registry = parsed as PlatformProfileRegistry;
  const profilesRaw = registry.profiles;
  if (!Array.isArray(profilesRaw) || profilesRaw.length === 0) {
    errors.push("profiles phải là mảng và có ít nhất một platform profile.");
  }

  const seenPlatforms = new Set<string>();
  const normalizedProfiles: PlatformProfile[] = [];

  if (Array.isArray(profilesRaw)) {
    profilesRaw.forEach((profileRaw, profileIndex) => {
      const profilePath = `profiles[${profileIndex}]`;
      if (!isRecord(profileRaw)) {
        errors.push(`${profilePath} phải là object.`);
        return;
      }

      const profile = profileRaw as PlatformProfile;
      const platform = normalizePlatformID(readString(profile.platform));
      const providerID = readString(profile.provider_id);
      const providerName = readString(profile.provider_name);
      const apiStyle = readString(profile.api_style);

      if (!platform) {
        errors.push(`${profilePath}.platform là bắt buộc.`);
      } else if (seenPlatforms.has(platform)) {
        errors.push(`platform bị trùng: ${platform}.`);
      } else {
        seenPlatforms.add(platform);
      }

      if (!isRecord(profile.guide)) {
        errors.push(`${profilePath}.guide là bắt buộc.`);
        normalizedProfiles.push({
          ...profile,
          platform,
          provider_id: providerID,
          provider_name: providerName,
          api_style: apiStyle,
        });
        return;
      }

      const guide = profile.guide as PlatformGuideMetadata;
      const title = readString(guide.title);
      const description = readString(guide.description);
      const profileID = readString(guide.profile_id) || platform;
      const docsURL = readString(guide.docs_url);
      const defaultClient = readString(guide.default_client);
      const clientsRaw = guide.clients;
      const copyBlocksRaw = guide.copy_blocks;

      if (!title) {
        errors.push(`${profilePath}.guide.title là bắt buộc.`);
      }
      if (!description) {
        errors.push(`${profilePath}.guide.description là bắt buộc.`);
      }

      const clientIDs = new Set<string>();
      const normalizedClients: PlatformGuideClient[] = [];
      if (clientsRaw !== undefined && !Array.isArray(clientsRaw)) {
        errors.push(`${profilePath}.guide.clients phải là mảng nếu được khai báo.`);
      }
      if (Array.isArray(clientsRaw)) {
        clientsRaw.forEach((clientRaw, clientIndex) => {
          const clientPath = `${profilePath}.guide.clients[${clientIndex}]`;
          if (!isRecord(clientRaw)) {
            errors.push(`${clientPath} phải là object.`);
            return;
          }
          const clientID = readString(clientRaw.id);
          const label = readString(clientRaw.label);
          if (!clientID || !label) {
            errors.push(`${clientPath}.id và ${clientPath}.label là bắt buộc.`);
          } else if (clientIDs.has(clientID)) {
            errors.push(`${clientPath}.id bị trùng: ${clientID}.`);
          } else {
            clientIDs.add(clientID);
          }
          normalizedClients.push({
            ...(clientRaw as PlatformGuideClient),
            id: clientID,
            label,
            os: readStringArray(clientRaw.os),
          });
        });
      }

      if (defaultClient && clientIDs.size > 0 && !clientIDs.has(defaultClient)) {
        warnings.push(
          `${profilePath}.guide.default_client (${defaultClient}) chưa có trong guide.clients.`,
        );
      }

      const blockIDs = new Set<string>();
      const normalizedBlocks: PlatformGuideCopyBlock[] = [];
      if (copyBlocksRaw !== undefined && !Array.isArray(copyBlocksRaw)) {
        errors.push(`${profilePath}.guide.copy_blocks phải là mảng nếu được khai báo.`);
      }
      if (Array.isArray(copyBlocksRaw)) {
        copyBlocksRaw.forEach((blockRaw, blockIndex) => {
          const blockPath = `${profilePath}.guide.copy_blocks[${blockIndex}]`;
          if (!isRecord(blockRaw)) {
            errors.push(`${blockPath} phải là object.`);
            return;
          }
          const blockID = readString(blockRaw.id);
          const clientID = readString(blockRaw.client_id);
          const path = readString(blockRaw.path);
          const contentTemplate = readString(blockRaw.content_template);
          if (!blockID || !clientID || !path || !contentTemplate) {
            errors.push(
              `${blockPath}.id, client_id, path, content_template là bắt buộc.`,
            );
          } else if (blockIDs.has(blockID)) {
            errors.push(`${blockPath}.id bị trùng: ${blockID}.`);
          } else {
            blockIDs.add(blockID);
          }
          if (clientID && clientIDs.size > 0 && !clientIDs.has(clientID)) {
            warnings.push(
              `${blockPath}.client_id (${clientID}) chưa có trong guide.clients.`,
            );
          }
          normalizedBlocks.push({
            ...(blockRaw as PlatformGuideCopyBlock),
            id: blockID,
            client_id: clientID,
            os: readString(blockRaw.os).toLowerCase(),
            path,
            hint: readString(blockRaw.hint),
            language: readString(blockRaw.language),
            content_template: typeof blockRaw.content_template === "string"
              ? blockRaw.content_template
              : contentTemplate,
          });
        });
      }

      normalizedProfiles.push({
        ...profile,
        platform,
        provider_id: providerID,
        provider_name: providerName,
        api_style: apiStyle,
        guide: {
          ...guide,
          profile_id: profileID,
          title,
          description,
          note: readString(guide.note),
          docs_url: docsURL,
          default_client: defaultClient,
          clients: normalizedClients,
          copy_blocks: normalizedBlocks,
        },
      });
    });
  }

  const missingDefaults = defaultPlatformIDs.filter(
    (platform) => !seenPlatforms.has(platform),
  );
  if (missingDefaults.length > 0) {
    warnings.push(
      `Registry đang thiếu default platform: ${missingDefaults.join(
        ", ",
      )}. Nếu cố ý chỉ override một phần thì backend vẫn lưu được, nhưng default baseline nên có đủ 3 nền tảng.`,
    );
  }

  const version = Number(registry.version || 1);
  const normalizedRegistry: PlatformProfileRegistry = {
    ...registry,
    version: Number.isFinite(version) && version > 0 ? version : 1,
    profiles: normalizedProfiles,
  };

  return {
    ok: errors.length === 0,
    isEmpty: false,
    registry: normalizedRegistry,
    errors,
    warnings,
  };
}

export function formatPlatformProfileRegistryInput(raw: string): string | null {
  const validation = validatePlatformProfileRegistryInput(raw);
  if (!validation.ok || !validation.registry) return null;
  return JSON.stringify(validation.registry, null, 2);
}

export function createPlatformProfileTemplate(platform: string): PlatformProfile {
  const normalizedPlatform = normalizePlatformID(platform) || "new-platform";
  const clientID = `${normalizedPlatform}-client`;
  return {
    platform: normalizedPlatform,
    provider_id: `${normalizedPlatform}-provider`,
    provider_name: normalizedPlatform,
    api_style: "openai-compatible",
    guide: {
      profile_id: normalizedPlatform,
      title: `${normalizedPlatform} delivery guide`,
      description: "Operator inserted guide metadata stored in platform_profile_registry.",
      note: "Edit clients and copy_blocks so the user modal can render this platform without hardcoded UI changes.",
      default_client: clientID,
      clients: [
        {
          id: clientID,
          label: `${normalizedPlatform} client`,
          os: ["unix", "windows"],
        },
      ],
      copy_blocks: [
        {
          id: `${normalizedPlatform}-config`,
          client_id: clientID,
          os: "unix",
          path: "Terminal",
          hint: "Default inserted block",
          language: "shell",
          content_template: `export API_BASE_URL="{{api_base_url}}"\nexport API_KEY="{{api_key}}"`,
        },
      ],
    },
  };
}

export function appendPlatformProfileTemplate(raw: string): {
  value: string;
  platform: string;
} | null {
  const trimmed = raw.trim();
  let root: PlatformProfileRegistry;
  if (!trimmed) {
    root = { version: 1, profiles: [] };
  } else {
    try {
      const parsed = JSON.parse(trimmed) as unknown;
      if (!isRecord(parsed)) return null;
      root = parsed as PlatformProfileRegistry;
    } catch (_error) {
      return null;
    }
  }

  const profiles = Array.isArray(root.profiles) ? [...root.profiles] : [];
  const existingPlatforms = new Set(
    profiles
      .map((profile) => normalizePlatformID(readString(profile.platform)))
      .filter(Boolean),
  );
  let nextPlatform = "new-platform";
  let suffix = 2;
  while (existingPlatforms.has(nextPlatform)) {
    nextPlatform = `new-platform-${suffix}`;
    suffix += 1;
  }

  profiles.push(createPlatformProfileTemplate(nextPlatform));
  const nextRegistry: PlatformProfileRegistry = {
    ...root,
    version: Number(root.version || 1),
    profiles,
  };

  return {
    value: JSON.stringify(nextRegistry, null, 2),
    platform: nextPlatform,
  };
}
